File: sorting-algos/merge_sort/mergeSort.java
import java.util.Scanner;

public class MergeSort {
   public static void main(String[] args) {
		Scanner s= new Scanner(System.in);
		System.out.print("Please enter the size of array to be sorted : \n");
		int n = s.nextInt();
		int arr[] = new int[n];
		for(int i=1;i<=n;i++){
			   System.out.print("Enter the "+ i + " element : " );	
			   arr[i-1] = s.nextInt();
		}
		int res[] = new int[n];
		arrDivide(arr,res,0,arr.length);
		System.out.println("Sorted Array is : ");
		for(int i=0;i<n;i++){
			   System.out.print(arr[i] + " "); 
		}
		s.close();
   }

/* In merge sort we first keep on dividing the array in two parts left and right 
   till we get each element separately then we start to merge the elements together
   sorting them and storing them in a separate array.
   Worst case time complexity for merge sort is nlogn.. 
*/
  
  public static void arrDivide(int arr[],int temp[],int left,int right){
	   
	   if(right-left==1 || left==right){     // when array length is 1 or 0 
		   return;			 // it is already sorted..	
	   }											
	   
 	   if(left<right){				
	   int mid = (left+right)/2;		 // dividing array into two parts using indices.
	   arrDivide(arr,temp,left,mid);	 // sorting first half.	
	   arrDivide(arr,temp,mid,right);	 // sorting other half.
	   mergeArrays(arr,temp,left,mid,right); // merging the two sorted halves. 
	   }

   }
   
   public static void mergeArrays(int arr[],int temp[],int left,int mid,int right){
      	   int i,j,k;
	   i = left; 
	   j=  mid;
	   k = left;
		
	   // merging the two divisions (left & right) in sorted order
	   // and storing in temp array.	
	   while(i!=mid && j!=right){		
		   if(arr[i]<=arr[j]){
			 temp[k++] = arr[i++];    
		   }else{
			 temp[k++] = arr[j++];  
		   }
	   }

	   // add the remaining elements.	
	   while(i!=mid){			
		   temp[k++] = arr[i++];
	   }
	   while(j!=right){
		   temp[k++] = arr[j++];
	   }
	   
	   for(i=left;i<right;i++){
		   arr[i] = temp[i];
	   }  
   }

}

/* Example:
INPUT: 
Please enter the size of array to be sorted : 
8
Enter the 1 element : 55
Enter the 2 element : 7
Enter the 3 element : 2
Enter the 4 element : 13
Enter the 5 element : 1
Enter the 6 element : 23
Enter the 7 element : 14
Enter the 8 element : 54

OUTPUT: 
Sorted Array is : 
1 2 7 13 14 23 54 55 
*/ 

File: sorting-algos/merge_sort/Merge_sort.cpp
#include <bits/stdc++.h>
using namespace std;
void Merge(int a[] , int start, int mid, int end)
{
	int i,j,k;
    int b=mid-start+1;
    int c=end-mid;
 
    int l[b], r[c]; //temp arrays
 
    for (i = 0; i < b; i++)
        l[i] = a[start + i];
    for (j = 0; j < c; j++)
        r[j] = a[mid + 1+ j];
 
    i=0;j=0;k=start;
    //Merge the 2 arrays in sorted order
    while (i < b && j < c)
    {
        if (l[i] <= r[j])
        {
            a[k] = l[i];
            i++;
        }
        else
        {
            a[k] = r[j];
            j++;
        }
        k++;
    }
 
    //add the remaining elements in sorted order
    while (i < b)
    {
        a[k] = l[i];
        i++;
        k++;
    }
 
    while (j < c)
    {
        a[k] = r[j];
        j++;
        k++;
    }

}
void MergeSort(int a[], int start , int end )
{
    if( start < end ) 
    {
     int mid = (start + end ) / 2 ;   //dividing the problem into sub-problems
     MergeSort(a,start,mid);      //Sort the first half
     MergeSort(a,mid+1,end);       //Sort the second half

     Merge(a,start,mid,end);  //Merge both the parts and sort them
}                    
}
int main() 
{
    int a[]= {55,7,2,13,1,23,14,54};
    int size = sizeof(a)/sizeof(a[0]);
    MergeSort(a,0,size-1);
    cout<<"Sorted Array is :\n";
    for(int i=0;i<size;i++)
	cout<<a[i]<<" ";
    return 0;
}

/*
Output : 
Sorted Array is :
1 2 7 13 14 23 54 55 
*/